File: app/src/main/java/com/college/portal/studentportal/data/LoginRepository.java
package com.college.portal.studentportal.data;

import com.college.portal.studentportal.callback.FirebaseAuthCompleteListener;
import com.college.portal.studentportal.data.model.LoggedInUser;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

/**
 * Class that requests authentication and user information from the remote data source and
 * maintains an in-memory cache of login status and user credentials information.
 */

public class LoginRepository {

    private static volatile LoginRepository instance;
    FirebaseAuth userAuth = FirebaseAuth.getInstance();
    FirebaseUser userDetails;
    private LoginDataSource dataSource;
    private static final String TAG = "Login";


    // If user credentials will be cached in local storage, it is recommended it be encrypted
    // @see https://developer.android.com/training/articles/keystore
    private LoggedInUser user = null;

    // private constructor : singleton access
    private LoginRepository(LoginDataSource dataSource) {
        this.dataSource = dataSource;
    }

    public static LoginRepository getInstance(LoginDataSource dataSource) {
        if (instance == null) {
            instance = new LoginRepository(dataSource);
        }
        return instance;
    }

    public boolean isLoggedIn() {
        return user != null;
    }

    public void logout() {
        user = null;
        dataSource.logout();
    }

    private void setLoggedInUser(LoggedInUser user) {
        this.user = user;
        // If user credentials will be cached in local storage, it is recommended it be encrypted
        // @see https://developer.android.com/training/articles/keystore
    }
    private Result<LoggedInUser> result;
    private LoggedInUser logged;
    public void login(String username, String password, FirebaseAuthCompleteListener listener) {
        // handle login
        userAuth.signInWithEmailAndPassword(username,password)
                .addOnSuccessListener(authResult -> {
                    userDetails = userAuth.getCurrentUser();
                    //TODO: REMOVE THE COMMENT AFTER UPLOADING USERS TO DATABASE FROM CONSOLE
                    //fetchLoggedInUserData(userDetails.getUid(),listener);
                    logged = new LoggedInUser(userDetails.getUid(),userDetails.getEmail(),"Admin","today","1");
                    result = new Result.Success<>(logged);
                    listener.onAuthComplete(result);
                })
                .addOnFailureListener(e -> {
                    Result.Error resultError = new Result.Error(e);
                    listener.onAuthFailure(resultError);
                });
    }

    private void fetchLoggedInUserData(String uid,FirebaseAuthCompleteListener listener){
        FirebaseFirestore.getInstance()
                .collection("users")
                .document(uid)
                .get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot documentSnapshot) {
                        logged = documentSnapshot.toObject(LoggedInUser.class);
                        result = new Result.Success<>(logged);
                        listener.onAuthComplete(result);
                    }
                });
    }
}

File: app/src/main/java/com/college/portal/studentportal/ui/login/LoginRepository.kt
package com.college.portal.studentportal.ui.login

import android.util.Log
import com.college.portal.studentportal.callback.FirebaseAuthCompleteListener
import com.college.portal.studentportal.data.model.LoggedInUser
import com.college.portal.studentportal.roomDatabase.user.CurrentUserDatabase
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Class that requests authentication and user information from the remote data source and
 * maintains an in-memory cache of login status and user credentials information.
 */
class LoginRepository {
    private var userAuth = FirebaseAuth.getInstance()
    private var userDetails: FirebaseUser? = null

    // If user credentials will be cached in local storage, it is recommended it be encrypted
    // @see https://developer.android.com/training/articles/keystore
    private var user: LoggedInUser? = null
    val isLoggedIn: Boolean
        get() = user != null

    fun logout() {
        user = null
    }

    private fun setLoggedInUser(user: LoggedInUser) {
        this.user = user
        // If user credentials will be cached in local storage, it is recommended it be encrypted
        // @see https://developer.android.com/training/articles/keystore
    }

    private var result: Result<LoggedInUser?>? = null
    private var logged: LoggedInUser? = null
    private lateinit var userDatabase: CurrentUserDatabase
    fun login(
        username: String?, password: String?,
        listener: FirebaseAuthCompleteListener,
        userDatabase: CurrentUserDatabase?
    ) {
        // handle login
        userAuth.signInWithEmailAndPassword(username!!, password!!)
            .addOnSuccessListener { authResult: AuthResult? ->

                userDetails = userAuth.currentUser
                this.userDatabase = userDatabase!!
                fetchLoggedInUserData(userDetails?.uid!!,listener)
            }
            .addOnFailureListener { e: Exception? ->
                val resultError = Result.Error(e)
                listener.onAuthFailure(resultError)
            }
    }

    private fun fetchLoggedInUserData(uid: String, listener: FirebaseAuthCompleteListener) {
        FirebaseFirestore.getInstance()
            .collection("users")
            .document(uid)
            .get()
            .addOnSuccessListener { documentSnapshot: DocumentSnapshot ->
                logged = documentSnapshot.toObject(
                    LoggedInUser::class.java
                )
                result = Result.Success<LoggedInUser?>(logged)

                GlobalScope.launch(Dispatchers.IO) {
                    val currentUserDao = userDatabase.getCurrentUserDao()
                    if (logged!=null){
                        currentUserDao.insertCurrentUser(logged!!)
                        withContext(Dispatchers.Main){
                            listener.onAuthComplete(result)
                        }
                    }else{
                        throw ClassCastException(
                            "Could not cast document snapshot data to CurrentUserEntity"
                        )
                    }
                }
            }
            .addOnFailureListener { e: Exception ->
                Log.i("user-data-fetch-failed", e.message, e)
            }
            .addOnCanceledListener {
                Log.d(
                    "user-data-fetch-canceled",
                    "May be canceled by the user itself"
                )
            }
    }
}